x-pack/filebeat/input/httpjson/request.go

```diff
@@ -146,38 +164,177 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+				if err != nil {
+					return err
+				}
+			}
+		} else {
+			if len(ids) == 0 {
+				n = 0
+				continue
+			}
+			urlCopy = rf.url
+			urlString = rf.url.String()
+			// perform request over collected ids
+			for _, id := range ids {
+				// reformat urls of requestFactory using ids
+				rf.url, err = generateNewUrl(rf.replace, urlString, []byte(id))
+				if err != nil {
+					return fmt.Errorf("failed to generate new URL: %w", err)
+				}
+
+				// collect data from new urls
+				httpResp, err = rf.collectResponse(stdCtx, trCtx, r, publisher)
+				if err != nil {
+					return fmt.Errorf("failed to execute rf.collectResponse: %w", err)
+				}
+				// store data according to response type
+				if rfSize == i+1 && len(ids) != 0 {
+					finalResps = append(finalResps, httpResp)
+				} else {
+					intermediateResps = append(intermediateResps, httpResp)
+				}
+			}
+			rf.url = urlCopy
+
+			if rfSize == i+1 {
+				split = r.responseProcessor.split
+				r.responseProcessor.split = nil
+				n, err = r.processAndPublishEvents(stdCtx, trCtx, publisher, finalResps, true)
+				if err != nil {
+					return err
+				}
+				r.responseProcessor.split = split
+				continue
+			} else {
+				ids, err = r.getIdsFromResponses(intermediateResps, r.requestFactories[i+1].replace)
+				if err != nil {
+					return err
+				}
+				split = r.responseProcessor.split
+				r.responseProcessor.split = nil
+				n, err = r.processAndPublishEvents(stdCtx, trCtx, publisher, intermediateResps, false)
+				if err != nil {
+					return err
+				}
+				r.responseProcessor.split = split
+			}
+		}
+	}
+
+	httpResp.Body.Close()
+	r.log.Infof("request finished: %d events published", n)
+
+	return nil
+}
+
+// getIdsFromResponses returns ids from responses
+func (r *requester) getIdsFromResponses(intermediateResps []*http.Response, replace string) ([]string, error) {
+	var b []byte
+	var ids []string
+	var err error
+	// collect ids from all responses
+	for _, resp := range intermediateResps {
+		if resp.Body != nil {
+			b, err = io.ReadAll(resp.Body)
+			if err != nil {
+				return nil, fmt.Errorf("error while reading response body: %v", err)
+			}
+		}
+
+		// get replace values from collected json
+		ids, err = parse(string(b), replace)
+		if err != nil {
+			return nil, fmt.Errorf("error while getting keys: %v", err)
+		}
 	}
-	defer httpResp.Body.Close()
+	return ids, nil
+}
 
-	eventsCh, err := r.responseProcessor.startProcessing(stdCtx, trCtx, httpResp)
+// processAndPublishEvents process and publish events based on response type
+func (r *requester) processAndPublishEvents(stdCtx context.Context, trCtx *transformContext, publisher inputcursor.Publisher, finalResps []*http.Response, publish bool) (int, error) {
+	eventsCh, err := r.responseProcessor.startProcessing(stdCtx, trCtx, finalResps)
 	if err != nil {
-		return err
+		return 0, fmt.Errorf("error recieving eventCh: %v", err)
```

Comment:
You should use `%w` here, also I believe we can make this error message more clear by describing what went wrong instead of saying what variable is not correctly populated.
```suggestion
		return 0, fmt.Errorf("error starting response processor: %w", err)
```

Reply:
Sure, I will update it.
Thank for comment